### PR TITLE
PEP 647: User-defined Type Guards

### PR DESCRIPTION
**Pull request by erictraut**
I've signed the CLA
bugs.python.org username: erictr

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@erictraut

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


## CLA Missing

Our records indicate the following people have not signed the CLA:

@msfterictraut

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: pep-0647.rst [hunk 1]
```diff
@@ -0,0 +1,328 @@
+PEP: 647
+Title: User-Defined Type Guards
+Version: $Revision$
+Last-Modified: $Date$
+Author: Eric Traut <erictr at microsoft.com>
+BDFL-Delegate: Guido van Rossum <guido@python.org>
+Discussions-To: Python-Dev <python-dev@python.org>
```
Comment (gvanrossum) [inline pep-0647.rst:7]:
```suggestion
Discussions-To: Python-Dev <typing-sig@python.org>
```
(Unless you want to have the discussion all over again.)

File: pep-0647.rst [hunk 2]
```diff
@@ -0,0 +1,328 @@
+PEP: 647
+Title: User-Defined Type Guards
+Version: $Revision$
+Last-Modified: $Date$
+Author: Eric Traut <erictr at microsoft.com>
+BDFL-Delegate: Guido van Rossum <guido@python.org>
```
Comment (gvanrossum) [inline pep-0647.rst:6]:
```suggestion
Sponsor: Guido van Rossum <guido@python.org>
```
(See PEP 1 for the difference.)

File: pep-0647.rst [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+Discussions-To: Python-Dev <python-dev@python.org>
+Status: Provisional
```
Comment (gvanrossum) [inline pep-0647.rst:8]:
```suggestion
Status: Draft
```

File: pep-0647.rst [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 07-Oct-2020
+Python-Version: 3.10
+Post-History: TBD
+Resolution: TBD
```
Comment (gvanrossum) [inline pep-0647.rst:14]:
```suggestion
```

File: pep-0647.rst [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+Last-Modified: $Date$
+Author: Eric Traut <erictr at microsoft.com>
+BDFL-Delegate: Guido van Rossum <guido@python.org>
+Discussions-To: Python-Dev <python-dev@python.org>
+Status: Provisional
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 07-Oct-2020
+Python-Version: 3.10
+Post-History: TBD
```
Comment (gvanrossum) [inline pep-0647.rst:13]:
```suggestion
Post-History: 28-Dec-2020
```

File: pep-0647.rst [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+Decorator Syntax
+----------------
+
+The use of a decorator was considered for defining type guards.
+
+::
+
+    @type_guard(List[str])
+    is_str_list(val: List[object]) -> bool: ...
```
Comment (gvanrossum) [inline pep-0647.rst:245]:
```suggestion
    def is_str_list(val: List[object]) -> bool: ...
```

**Review by gvanrossum — COMMENTED** [review]
A few nits.

**Comment by gvanrossum** [comment]
Looks like the CLA bot needs to know about both your GitHub usernames.

Possibly there's a corporate CLA, but I don't know anything about it (I'll ask Brett).

**Comment by erictraut** [comment]
I initially registered with the wrong GitHub user name. I changed it to the one I used for the PR, but I think I did so too late. Not sure how to make the bot happy at this point. Apologies for the mix-up.

**Comment by gvanrossum** [comment]
There's a human in the loop (and she may be OOF for the holidays). I trust you signed the CLA so I will merge regardless.

**Pull request merged**